### PR TITLE
Support different layouts in imgcodec's Convert

### PR DESCRIPTION
**Pull request by szkarpinski**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:

Imgcodec's Convert was assuming channel-last layout of processed tensors. This PR lifts this assumption, allowing for any layout.

Supporting layout transposition (i.e. changing layout in Convert) is out of scope of this PR.

### Affected modules and functionalities:

Apart from changes to `Convert`, the following changes were made:
- In `decoder_test_helper.h`, `Crop` utility was adjusted to support layouts other than channel-last.
- In `convert_test.cc` a new class (`ConvertLayoutTest`) was added. Code shared by `ConvertLayoutTest` and old `ColorConversionTest` was extracted to `ConversionTestBase`.

### Tests:
The tests use data from https://github.com/NVIDIA/DALI_extra/pull/96

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2953

### CODE REVIEW

File: dali/imgcodec/util/convert_test.cc [hunk 1]
```diff
@@ -220,6 +235,57 @@ TYPED_TEST(ColorConversionTest, BgrToBgr) {
...
+    auto rgb_path = GetPath(layout_name, "rgb");
+    auto ycbcr_path = GetPath(layout_name, "ycbcr");
+
+    std::string layout_code = layout_name;
+    for (auto &c : layout_code) c = toupper(c);
+    TensorLayout layout(layout_code);
+
+    auto ref = ReadReferenceFrom(ycbcr_path);
+    if (roi) {
+      ref.SetLayout(layout);
```
Comment (klecki) [inline dali/imgcodec/util/convert_test.cc:253]: Why do we set layout only when roi is specified?

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 2]
```diff
@@ -303,14 +303,18 @@ class DecoderTestBase : public ::testing::Test {
...
+
     int ndim = shape.sample_dim();
     if (roi.begin.size() == ndim - 1) {
-      roi.begin = shape_cat(roi.begin, 0);
+      roi.begin = shape_cat(shape_cat(roi.begin.first(channel_dim), 0),
+                            roi.begin.last(roi.begin.size() - channel_dim));
     }
     if (roi.end.size() == ndim - 1) {
-      roi.end = shape_cat(roi.end, *(shape.end() - 1));
+            roi.end = shape_cat(shape_cat(roi.end.first(channel_dim), shape[channel_dim]),
```
[the remark this answers is not in the document]
Comment (szkarpinski) [reply dali/imgcodec/decoders/decoder_test_helper.h]: Fixed

File: dali/imgcodec/util/convert_test.cc (hunk 1 above)
Comment (szkarpinski) [reply dali/imgcodec/util/convert_test.cc]: The layout is only needed by `Crop` and that's how it ended up here. I agree that keeping it inside the `if` is confusing, I moved it out

**Review by klecki — APPROVED** [review]

**Review by jantonguirao — APPROVED** [review]

**Comment by szkarpinski** [comment]
!build

**Pull request merged**